drivers/axi_core/axi_dmac/axi_dmac.c

```diff
@@ -123,204 +255,290 @@ int32_t axi_dmac_write(struct axi_dmac *dmac,
-	uint32_t timeout = 0;
-
-	if (size == 0)
-		return 0; /* nothing to do */
+	uint32_t reg_val, burst_size;
 
-	axi_dmac_write(dmac, AXI_DMAC_REG_CTRL, 0x0);
-	axi_dmac_write(dmac, AXI_DMAC_REG_CTRL, AXI_DMAC_CTRL_ENABLE);
+	if (dma_transfer->size == 0)
+		return 0; /* Nothing to do. */
 
-	axi_dmac_write(dmac, AXI_DMAC_REG_IRQ_MASK, 0x0);
+	/* Set current transfer parameters. */
+	dmac->transfer.size = dma_transfer->size;
+	dmac->transfer.cyclic = dma_transfer->cyclic;
+	dmac->transfer.dest_addr = dma_transfer->dest_addr;
+	dmac->transfer.src_addr = dma_transfer->src_addr;
 
-	axi_dmac_read(dmac, AXI_DMAC_REG_TRANSFER_ID, &transfer_id);
-	axi_dmac_read(dmac, AXI_DMAC_REG_IRQ_PENDING, &reg_val);
-	axi_dmac_write(dmac, AXI_DMAC_REG_IRQ_PENDING, reg_val);
+	dmac->remaining_size = dma_transfer->size;
+	dmac->next_dest_addr = dma_transfer->dest_addr;
+	dmac->next_src_addr = dma_transfer->src_addr;
 
-	switch (dmac->direction) {
-	case DMA_DEV_TO_MEM:
-		axi_dmac_write(dmac, AXI_DMAC_REG_DEST_ADDRESS, address);
-		axi_dmac_write(dmac, AXI_DMAC_REG_DEST_STRIDE, 0x0);
-		break;
-	case DMA_MEM_TO_DEV:
-		axi_dmac_write(dmac, AXI_DMAC_REG_SRC_ADDRESS, address);
-		axi_dmac_write(dmac, AXI_DMAC_REG_SRC_STRIDE, 0x0);
-		break;
-	default:
-		return -1; // Other directions are not supported yet
+	/* If HW cyclic transfer selected and not available, show error */
+	/* HW cyclic transfer available only for MEM to DEV transfers. */
+	if ((!dmac->hw_cyclic) && (dma_transfer->cyclic == CYCLIC)) {
+		printf("Transfer mode not supported!\n");
+		return -1;
 	}
 
-	if ((size - 1) > dmac->transfer_max_size) {
-		dmac->big_transfer.address = address;
-		dmac->big_transfer.size = size - 1;
-		dmac->big_transfer.size_done = dmac->transfer_max_size;
-
-		axi_dmac_write(dmac, AXI_DMAC_REG_X_LENGTH,
-			       dmac->transfer_max_size);
-		axi_dmac_write(dmac, AXI_DMAC_REG_Y_LENGTH, 0x0);
-
-		axi_dmac_write(dmac, AXI_DMAC_REG_FLAGS, dmac->flags);
-
-		axi_dmac_write(dmac, AXI_DMAC_REG_START_TRANSFER, 0x1);
-
-		while(!dmac->big_transfer.transfer_done) {
-			timeout++;
-			if (timeout == UINT32_MAX)
-				return -1;
+	/* Cyclic transfers not possible for DEV_TO_MEM and MEM_TO_MEM transmissions. */
+	if ((dmac->direction == DMA_DEV_TO_MEM)
+	    || (dmac->direction == DMA_MEM_TO_MEM)) {
+		if (dma_transfer->cyclic == CYCLIC) {
+			printf("Transfer mode not supported!\n");
+			return -1;
 		}
-
-		dmac->big_transfer.address = 0;
-		dmac->big_transfer.size = 0;
-		dmac->big_transfer.size_done = 0;
-
-		return 0;
 	}
 
-	axi_dmac_write(dmac, AXI_DMAC_REG_X_LENGTH, size - 1);
-	axi_dmac_write(dmac, AXI_DMAC_REG_Y_LENGTH, 0x0);
-
-	axi_dmac_write(dmac, AXI_DMAC_REG_FLAGS, dmac->flags);
+	/* Cyclic transfers set to HW or SW depending on size for MEM to DEV. */
+	if ((dmac->direction == DMA_MEM_TO_DEV) && (dmac->transfer.cyclic == CYCLIC)) {
+		if ((dmac->remaining_size - 1) <= dmac->max_length) {
+			if (dmac->hw_cyclic)
+				axi_dmac_write(dmac, AXI_DMAC_REG_FLAGS, DMA_CYCLIC);
```

Comment:
We could do a read-modify-write here as well to keep other flags if necessary.

Reply:
You are right, I will modify the code accordingly.